draft-ietf-quic-tls.md

```diff
@@ -255,142 +205,169 @@ A simplified TLS 1.3 handshake with 0-RTT application data is shown in
-transitions from no protection (@H) to full 1-RTT protection (@1) after it sends
-the last of its handshake messages.
-
-Some TLS handshake messages are protected by the TLS handshake record
-protection.  These keys are not exported from the TLS connection for use in
-QUIC.  QUIC packets from the server are sent in the clear until the final
-transition to 1-RTT keys.
-
-The client transitions from handshake (@H) to 0-RTT keys (@0) when sending 0-RTT
-data, and subsequently to to 1-RTT keys (@1) after its second flight of TLS
-handshake messages.  This creates the potential for unprotected packets to be
-received by a server in close proximity to packets that are protected with 1-RTT
-keys.
-
-More information on key transitions is included in {{hs-protection}}.
+{: #schematic title="QUIC and TLS Interactions"}
 
+Unlike TLS over TCP, QUIC applications which want to send data do not
+send it through TLS "application_data" records. Rather, they send it
+as QUIC STREAM frames which are then carried in QUIC packets.
+
+
+# Carrying TLS Messages {#carrying-tls}
+
+QUIC carries TLS handshake data in CRYPTO_HS frames, each of which
+consists of a contiguous block of handshake data identified by an
+offset and length. Those frames are packaged into QUIC packets
+and encrypted under the current TLS encryption level.
+As with TLS over TCP, once TLS handshake data has
+been delivered to QUIC, it is QUIC's responsibility to deliver it
+reliably. Each chunk of data is associated with the then-current TLS
+sending keys, and if QUIC needs to retransmit that data, it MUST use
+the same keys even if TLS has already updated to newer keys.
+
+One important difference between TLS 1.3 records (used with TCP)
+and QUIC CRYPTO_HS frames is that in QUIC multiple frames may appear
+in the same QUIC packet as long as they are associated with the
+same encryption level. For instance, an implementation might
+bundle a Handshake message and an ACK for some Handshake
+data into the same packet.
+
+In general, the rules for which data can appear in packets of which
+encryption level are the same in QUIC as in TLS over TCP:
+
+- CRYPTO_HS frames MAY appear in packets of any encryption level.
+- CONNECTION_CLOSE and CRYPTO_CLOSE MAY appear in packets of any
+  encryption level other than 0-RTT.
+- PADDING and PING frames MAY appear in packets of any encryption level.
+- ACK frames MAY appear in packets of any encryption level, but
+  MUST only acknowledge packets which appeared in that encryption
+  level.
+- STREAM frames MUST ONLY appear in the 0-RTT and 1-RTT levels.
+- All other frame types MUST only appear at the 1-RTT levels.
+
+Because packets may be reordered on the wire, QUIC uses the packet
+type to indicate which level a given packet was encrypted
+under, as shown in {{packet-types-levels}}. When multiple packets of
+different encryption levels need to be sent, endpoints SHOULD use
+coalesced packets to send them in the same UDP datagram.
```

Comment:
I think that, up to this point, we were relying on the fact that typical UDP stacks give us a clear per-packet boundary and this changes that.  Do we now see implementations receiving a UDP packet and assuming it might contain more than one QUIC packet?
It seems like the only way to implement this is to start parsing the QUIC packet and then, if there are leftover bytes, interpret that as another QUIC packet.
It's not too hard but it seems like a significant change.

Reply:
That's already documented under coalesced packets, I think: https://tools.ietf.org/html/draft-ietf-quic-transport-12#section-4.6

Reply:
OK, got it.